File: app/src/main/java/org/neidhardt/dynamicsoundboard/navigationdrawer/header/views/NavigationDrawerHeaderPresenter.java
package org.neidhardt.dynamicsoundboard.navigationdrawer.header.views;

import android.support.annotation.NonNull;
import org.neidhardt.dynamicsoundboard.navigationdrawer.header.events.OpenSoundLayoutsEvent;
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundlayouts.events.*;
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.views.SoundLayoutSettingsDialog;
import org.neidhardt.dynamicsoundboard.presenter.BaseViewPresenter;
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.model.SoundLayoutsAccess;

/**
 * File created by eric.neidhardt on 27.05.2015.
 */
public class NavigationDrawerHeaderPresenter
		extends
			BaseViewPresenter<NavigationDrawerHeader>
		implements
			SoundLayoutSettingsDialog.OnSoundLayoutRenamedEventListener,
			OnSoundLayoutRemovedEventListener,
			OnSoundLayoutSelectedEventListener
{
	private static final String TAG = NavigationDrawerHeaderPresenter.class.getName();

	private SoundLayoutsAccess soundLayoutModel;

	public void setSoundLayoutModel(SoundLayoutsAccess soundLayoutModel)
	{
		this.soundLayoutModel = soundLayoutModel;
	}

	@Override
	protected boolean isEventBusSubscriber()
	{
		return true;
	}

	@Override
	public void onAttachedToWindow()
	{
		super.onAttachedToWindow();
		if (this.getView() == null)
			throw new NullPointerException(TAG +": supplied view is null");

		if (this.soundLayoutModel == null)
			throw new NullPointerException(TAG +": supplied model is null");

		this.getView().showCurrentLayoutName(this.soundLayoutModel.getActiveSoundLayout().getLabel());
	}

	@Override
	@SuppressWarnings("unused")
	public void onEvent(SoundLayoutRenamedEvent event)
	{
		if (this.getView() == null || this.soundLayoutModel == null)
			return;

		this.getView().showCurrentLayoutName(this.soundLayoutModel.getActiveSoundLayout().getLabel());
	}

	@Override
	@SuppressWarnings("unused")
	public void onEvent(@NonNull SoundLayoutRemovedEvent event)
	{
		if (this.getView() == null || this.soundLayoutModel == null)
			return;

		this.getView().showCurrentLayoutName(this.soundLayoutModel.getActiveSoundLayout().getLabel());
	}

	@Override
	@SuppressWarnings("unused")
	public void onEvent(@NonNull SoundLayoutSelectedEvent event)
	{
		if (this.getView() == null || this.soundLayoutModel == null)
			return;

		this.getView().animateLayoutChanges();
		this.getView().showCurrentLayoutName(this.soundLayoutModel.getActiveSoundLayout().getLabel());
	}

	public void onChangeLayoutClicked()
	{
		this.getView().animateLayoutChanges();
		this.getEventBus().post(new OpenSoundLayoutsEvent());
	}

	public interface OnOpenSoundLayoutsEvent
	{
		/**
		 * This is called by greenRobot EventBus when the user clicks on the open SoundLayouts button in navigation drawer header.
		 * @param event delivered OpenSoundLayoutsEvent
		 */
		void onEvent(OpenSoundLayoutsEvent event);
	}
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/navigationdrawer/header/views/NavigationDrawerHeaderPresenter.kt
package org.neidhardt.dynamicsoundboard.navigationdrawer.header.views

import org.neidhardt.dynamicsoundboard.navigationdrawer.header.events.OpenSoundLayoutsRequestedEvent
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundlayouts.events.*
import org.neidhardt.dynamicsoundboard.presenter.BaseViewPresenter
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.model.SoundLayoutsAccess
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.views.SoundLayoutSettingsDialog

/**
 * File created by eric.neidhardt on 27.05.2015.
 */
class NavigationDrawerHeaderPresenter(private val soundLayoutModel: SoundLayoutsAccess?) :
		BaseViewPresenter<NavigationDrawerHeader>(),
		SoundLayoutSettingsDialog.OnSoundLayoutRenamedEventListener,
		OnSoundLayoutRemovedEventListener,
		OnSoundLayoutSelectedEventListener
{
	private val TAG = javaClass.name

	override fun isEventBusSubscriber(): Boolean = true

	override fun onAttachedToWindow() {
		super.onAttachedToWindow()
		if (this.view == null)
			throw NullPointerException(TAG + ": supplied view is null")

		if (this.soundLayoutModel == null)
			throw NullPointerException(TAG + ": supplied model is null")

		this.view.showCurrentLayoutName(this.soundLayoutModel.getActiveSoundLayout().label)
	}

	@SuppressWarnings("unused")
	override fun onEvent(event: SoundLayoutRenamedEvent) {
		if (this.view == null || this.soundLayoutModel == null)
			return

		this.view.showCurrentLayoutName(this.soundLayoutModel.getActiveSoundLayout().label)
	}

	@SuppressWarnings("unused")
	override fun onEvent(event: SoundLayoutRemovedEvent) {
		if (this.view == null || this.soundLayoutModel == null)
			return

		this.view.showCurrentLayoutName(this.soundLayoutModel.getActiveSoundLayout().label)
	}

	@SuppressWarnings("unused")
	override fun onEvent(event: SoundLayoutSelectedEvent) {
		if (this.view == null || this.soundLayoutModel == null)
			return

		this.view.animateLayoutChanges()
		this.view.showCurrentLayoutName(this.soundLayoutModel.getActiveSoundLayout().label)
	}

	fun onChangeLayoutClicked() {
		this.view.animateLayoutChanges()
		this.eventBus.post(OpenSoundLayoutsRequestedEvent())
	}
}
